Assets/Scripts/Models/Power/Grid.cs

```diff
@@ -118,28 +118,45 @@ private void ChargeAccumulators(ref float currentPowerLevel)
         {
             foreach (Connection connection in connections.Where(connection => connection.IsPowerAccumulator && !connection.IsFull))
             {
-                if (currentPowerLevel - connection.InputRate < 0.0f)
+                float inputRate = connection.AccumulatedPower + connection.InputRate > connection.Capacity ?
+                    connection.Capacity - connection.AccumulatedPower :
+                    connection.InputRate;
+
+                if (currentPowerLevel - inputRate < 0.0f)
                 {
-                    continue;
+                    inputRate = currentPowerLevel;
                 }
 
-                currentPowerLevel -= connection.InputRate;
-                connection.AccumulatedPower += connection.InputRate;
+                currentPowerLevel -= inputRate;
+                connection.AccumulatedPower += inputRate;
+
+                if (currentPowerLevel.IsZero())
+                {
+                    break;
+                }
             }
         }
 
         private void DischargeAccumulators(ref float currentPowerLevel)
         {
             if (currentPowerLevel +
-                connections.Where(connection => connection.IsPowerAccumulator && !connection.IsEmpty).Sum(powerRelated => powerRelated.OutputRate) < 0)
+                connections.Where(connection => connection.IsPowerAccumulator && !connection.IsEmpty)
+                .Sum(connection =>
+                connection.AccumulatedPower - connection.OutputRate < 0.0f ?
+                connection.AccumulatedPower :
+                connection.OutputRate) < 0)
```

Comment:
Ive been staring at those four lines for a while now (144-147), something needs to happen but Im not sure what. The expression is very complicated. Could we indent 144 with one tab and 145, 146, 147 with two tabs? Or, consider breaking out the logic into methods so that you just do `.Sum(MyMethod);`.